Hybrid and Prefix Printer

tkutcher opened the pull request:
Hi - new to Flutter/Dart but this has been a helpful package. This pull request would add two `LogPrinter` decorators.

1) `PrefixPrinter`  in prefix_printer.dart
This allows you to wrap any logger with a prefix to all of it's logs. Inspiration for this being #2 and the AndroidStudio grep-console plugin. But this could also be used just in general to decorate any logger:

```dart
final logger = Logger(
  printer:  PrefixPrinter(PrettyPrinter())
);
```
for example would add prefixes (`"DEBUG"`, `"INFO"`, etc.) to each line in the `PrettyPrinter` output. But since it is a decorator you could use the same thing for any other printer.

It also has optional params for each log level to specify the prefix:
```dart
PrefixPrinter(SimplePrinter(), debug: "DBUG", error: "OOPS");
```

Light unit tests in prefix_printer_test.dart

---

2) `HybridPrinter` in hybrid_printer.dart
This allows you to use a different printer for specific log levels. Inspiration for this is that I found with my logs if they were debug logs, I just wanted a simple one-line view of the log, and with regular logs the full `PrettyPrinter` log. With the `HybridPrinter` you can accomplish that like:

```dart
final logger = Logger(
  printer:  HybridPrinter(PrettyPrinter(), debug: SimplePrinter()),
  level: config.logLevel
);
```

Same thing as above where there are parameters for each level. 

Light unit tests in hybrid_printer_test.dart

---

So, just for fun, using them together (how I currently am):

```dart
final logger = Logger(
  printer:  PrefixPrinter(
      HybridPrinter(
          PrettyPrinter(
            methodCount: 2,
            colors: false,
            printTime: true
          ),
        debug: SimplePrinter()
      )),//OneLinePrinter(),
  level: config.logLevel
);
```

![image](https://user-images.githubusercontent.com/25716930/81469673-584fff80-91b4-11ea-83f5-a33f05573acf.png)

The workaround for #2 thus becomes turning colors off, and using prefixes you want for the grep console plugin. Not a full fix, but works nicely for me!

---

Thought I'd submit this MR in hopes you wanted to add it to the project (also wasn't sure to submit it to master or another branch). Project is awesome! I think these could help with more flexibility in people's own loggers. In general, I would be happy to contribute.

Cheers,

Tim

haarts commented on lib/src/printers/prefix_printer.dart:
```diff
@@ -0,0 +1,29 @@
+  Map<Level, String> _prefixMap;
+
+  PrefixPrinter(this._realPrinter,
+      {debug, verbose, wtf, info, warning, error, nothing}) : super() {
```
You can leave the `super()` out.

haarts commented on lib/src/printers/hybrid_printer.dart:
```diff
@@ -0,0 +1,38 @@
+  var _printerMap;
+
+  HybridPrinter(this._realPrinter,
+      {debug, verbose, wtf, info, warning, error, nothing}) : super() {
```
No need for the `super()` call here.

haarts commented on lib/src/printers/prefix_printer.dart:
```diff
@@ -0,0 +1,29 @@
+  PrefixPrinter(this._realPrinter,
+      {debug, verbose, wtf, info, warning, error, nothing}) : super() {
+    _prefixMap = {
+      Level.debug: debug ?? '  DEBUG ',
```
What's with the leading and trailing white space? At first I thought it was to align text but the strings vary in length.

tkutcher replied:
Yea the idea was to line them up (not dart-savvy enough to easily do string formatting and was lazy) - I suppose `Level.NOTHING` doesn't line up but I guess even specifying it in the first place is redundant. Smarter fix is probably to dynamically set the padded length of the string to be the max of them all. That way if you pass in a custom message to the constructor you don't have to make sure it lines up with everything else. I should update that.

tkutcher replied:
Updated this to dynamically pad strings to the length of the longest prefix.

haarts commented on test/printers/hybrid_printer_test.dart:
```diff
@@ -0,0 +1,46 @@
+    latestEvent = event;
+    return realPrinter.log(event);
+  }
+}
```
Neat! Well done.

haarts commented:
Ow! I would like for you to add your workaround for the colorless IDE terms to the README as to help others.

tkutcher commented:
Sounds good, thanks for the feedback! Later this evening I'll clean up those things, squash the cleanup commits, and add to the README for it on this branch.

-Tim

tkutcher commented:
Cleaned up and squashed relevant commits. So now there is one commit for the source code, and one for the note in the README.

Thanks and let me know if there's anything else!

haarts commented:
Well done man! Thanks for you contribution. I'll be releasing a new version with this shortly.
